File: solucion.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

// struct to store the given parameters of the paginated document

type PaginatedParameters struct {
	MaxLineSize  int
	MaxLineCount int
}

// Function to check if there is an error and panic if there is one

func checkError(e error) {
	if e != nil {
		panic(e)
	}
}

// Function to open the input document

func openInputDocument(path string) *os.File {
	f, err := os.Open(path)
	checkError(err)
	return f
}

// Function to create the output document

func createOutputDocument(path string) *os.File {
	out, err := os.Create(path)
	checkError(err)
	return out
}

// Function to paginate the input document and write on the output document

func paginateDocument(out *os.File, scanner *bufio.Scanner, params PaginatedParameters) {

	var line string              // variable to store the line to be written
	var lineCount, pageCount int // variables to store the line count and page count

	pageCount = 1 // set the page count to 1

	for scanner.Scan() {
		word := scanner.Text() // read the word from the input document

		if len(line)+len(word)+1 > params.MaxLineSize { // check if the line size + the next word size + next space exceeds the maximum line size
			_, err := fmt.Fprintln(out, line) // write the line on the output document
			checkError(err)

			line = word // set the line to the next word
			lineCount++

			if lineCount == params.MaxLineCount { // check if the line count is equal to the maximum line count
				_, err := fmt.Fprintf(out, "=== Page %d ===\n", pageCount)
				checkError(err)
				pageCount++
				lineCount = 0
			}

		} else { // else add a space if the line is not empty and add the word to the line
			if len(line) > 0 {
				line += " "
			}
			line += word
		}
	}

	if len(line) > 0 { // check if the last line is not empty
		_, err := fmt.Fprintln(out, line) // write the last line on the output document
		checkError(err)
		lineCount++
	}

	if lineCount > 0 { // check if the line count is greater than 0
		_, err := fmt.Fprintf(out, "=== Page %d ===", pageCount) // write the last page count on the output document
		checkError(err)
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
}

func main() {

	params := PaginatedParameters{
		MaxLineSize:  80,
		MaxLineCount: 25,
	}

	var inputDocument = "./document.txt" // declare a variable to store the name of the input document route

	var f = openInputDocument(inputDocument) // create a file object to read the input document

	var outputDocument = "./paginated_document_go.txt" // declare a variable to store the name of the output document route

	var out = createOutputDocument(outputDocument) // create a file object to write on the output document

	scanner := bufio.NewScanner(f) // create a scanner object to read the input document
	scanner.Split(bufio.ScanWords) // set the scanner to read word by word

	paginateDocument(out, scanner, params)
}

File: solucion.cc
#include <iostream>
#include <fstream>
#include <string>
using namespace std;

// create a struc to store the initial parameters

struct PaginatedParameters {
    int maxLineSize = 80;
    int maxLineCount = 25;
};

// check if the ifstream file was opened correctly

void checkError(ifstream& f) {
    if (!f) {
        cout << "Error opening input file" << endl;
        exit(1);
    }

}

// check if the ofstream file was opened correctly

void checkError(ofstream& out) {
    if (!out) {
        cout << "Error opening output file" << endl;
        exit(1);
    }
}

// Function to open the input document, checking if it was opened correctly

void openInputDocument(ifstream& f, string& inputDocumentName) {
    f.open(inputDocumentName);
    checkError(f);
}

// Function to open the output document, checkign if it was opened correctly

void openOutputDocument(ofstream& out, string& outputDocumentName) {
    out.open(outputDocumentName);
    checkError(out);
}

void paginateDocument (ifstream& f, ofstream& out, int& maxLineSize, int& maxLineCount, string& word, string& line, int& lineCount, int& pageCount) {

    // read the document.txt file word by word

    while (f >> word){
        if (line.size() + word.size() + 1 > 80){ // if the line + the next word + a space is bigger than maxLineSize characters
            out << line << endl; //print the line as it was
            line = word; // start the next one with that word
            lineCount++;
            if (lineCount == 25){ // if the line count is 25, print the page number and reset the line count
                out << "=== Page " << pageCount << " ===" << endl;
                pageCount++;
                lineCount = 0;
            }
        } else {
            line += (line.empty() ? "" : " ") + word; // if the line + the next word is smaller than 80 characters, add the word to the line, separated by a space if needed (there was a word before)
        }
    }

    if (!line.empty()) { // if there is a line left to print, print it
        out << line << endl;
        lineCount++;
    }
    if (lineCount > 0) { // if the line count is bigger than 0, print the last page number
        out << "=== Page " << pageCount << " ===" << endl; 
    }
}

int main() {
    PaginatedParameters paginatedParameters;

    ifstream f; 
    string inputDocumentName = "document.txt"; // declare a string to store the document name
    openInputDocument(f, inputDocumentName); // open the document.txt file

    ofstream out;
    string outputDocumentName = "paginated_document_cc.txt"; // declare a string to store the solution name
    openOutputDocument(out, outputDocumentName); // open / create the paginated_document.txt file where the solution will be stored

    string word = ""; // declare a strings to store each word of the document
    string line = ""; // declare a string to store each line of the document
    
    int lineCount = 0; // declare an int to count the line number
    int pageCount = 1; // declare an int to count the page number (starts at 1)


    paginateDocument(f, out, paginatedParameters.maxLineSize, paginatedParameters.maxLineCount, word, line, lineCount, pageCount); // call the solution function

    f.close();
    out.close();
    return 0;
}